cronback-services/src/api/auth_middleware.rs

```diff
@@ -1,17 +1,34 @@
 use std::sync::Arc;
 
-use axum::extract::State;
+use axum::extract::{FromRef, State};
 use axum::http::{self, HeaderMap, HeaderValue, Request};
 use axum::middleware::Next;
 use axum::response::IntoResponse;
 use lib::prelude::*;
 
-use super::auth::{AuthError, SecretApiKey};
+use super::auth::{AuthError, Authenticator, SecretApiKey};
 use super::errors::ApiError;
 use super::AppState;
 
 const ON_BEHALF_OF_HEADER_NAME: &str = "X-On-Behalf-Of";
 
+// Partial state from the main app state to facilitate writing tests for the
+// middleware.
+#[derive(Clone)]
+pub struct AuthenticationState {
+    authenticator: Authenticator,
+    config: super::config::ApiSvcConfig,
+}
+
+impl FromRef<Arc<AppState>> for AuthenticationState {
+    fn from_ref(input: &Arc<AppState>) -> Self {
+        Self {
+            authenticator: input.authenticator.clone(),
+            config: input.context.service_config(),
+        }
+    }
+}
```

Comment:
I thought FromRef was implemented for all `T: Clone` already, no? I wonder why this wasn't the case for `AuthenticationState`.

Also, did you try to `#[derive(FromRef, Clone)]`?

Reply:
`FromRef<T> for T` is implemented for all Ts where `T: Clone` but here I'm defining how to get an `AuthenticationState` (custom struct) from an `AppState`, there's no way the compiler can infer this on its own or even with a derive macro, right?

I think the derive macro might work only if I'm getting a state member from the state struct, but here it's a completely custom new struct.